### PR TITLE
Add test for documentation

### PR DESCRIPTION
**Pull request by Dref360**
Added a test to enforce doc in the core modules.

I leave legacy behind since it's deprecated and the from_config function since it's only used internally.

### CODE REVIEW

File: tests/test_documentation.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
+from inspect import isfunction, isclass, getmembers, ismodule, ismethod, signature, getmodule
```
Comment (fchollet) [inline tests/test_documentation.py]: Prefer `import inspect` then use e.g. `inspect.isfunction`

File: tests/test_documentation.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+modules = ['keras.layers', 'keras.models', 'keras', 'keras.backend.tensorflow_backend']
+accepted_name = ["from_config"]
+accepted_module = ["keras.legacy.layers", "keras.utils.generic_utils"]
+
+
+def handle_class(name, member):
+    if name in accepted_name or member.__module__ in accepted_module:
+        return
+    assert member.__doc__ is not None, "class doesn't have any documentation {} {}".format(name, member.__module__, getmodule(member).__file__)
```
Comment (fchollet) [inline tests/test_documentation.py]: Lines are in general too long, limit line length to something reasonable (e.g. 80 characters).

Use `Exception` subclasses (e.g. `ValueError`) instead of `assert`.

File: keras/backend/tensorflow_backend.py [hunk 3]
```diff
@@ -43,6 +43,14 @@
 
 
 def get_uid(prefix=''):
+    """Get the uid for the default graph.
+
+    #Arguments
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: These section titles are formatted with `#` followed by a space (add space).

File: keras/backend/tensorflow_backend.py [hunk 4]
```diff
@@ -169,6 +178,13 @@ def set_session(session):
 # VARIABLE MANIPULATION
 
 def _convert_string_dtype(dtype):
+    """
+    Get the type from a string.
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: The one-line description should be on the same line as `"""`

File: keras/backend/tensorflow_backend.py [hunk 5]
```diff
@@ -169,6 +178,13 @@ def set_session(session):
...
+    # Arguments
+        dtype: A string representation of a type.
+    # Returns:
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: The sections should be separated with a newline (add newline).

File: keras/backend/tensorflow_backend.py [hunk 6]
```diff
@@ -309,6 +333,16 @@ def _initialize_variables():
 
 
 def constant(value, dtype=None, shape=None, name=None):
+    """Create a constant tensor.
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: We typically conjugate the starting verb in the docstrings (e.g. `creates`).

File: tests/test_documentation.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
+from inspect import isfunction, isclass, getmembers, ismodule, ismethod, signature, getmodule
+import importlib
+from itertools import compress
+import pytest
+
+from keras.utils.test_utils import keras_test
+
+modules = ['keras.layers', 'keras.models', 'keras', 'keras.backend.tensorflow_backend']
+accepted_name = ["from_config"]
```
Comment (fchollet) [inline tests/test_documentation.py]: Don't mix `"` and `'`, normalize to `'`.

File: tests/test_documentation.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+        if ismethod(met):
+            handle_method(n, met)
+
+
+def handle_function(name, member):
+    if name in accepted_name or member.__module__ in accepted_module:
+        return
+    doc = member.__doc__
+    assert doc is not None, "{} function doesn't have any documentation {} {}".format(name, member.__module__, getmodule(member).__file__)
+    args = list(signature(member).parameters.keys())
```
Comment (fchollet) [inline tests/test_documentation.py]: We should not only test for argument presence, we should test for:

- section presence, e.g. `# Returns`, IFF the function returns something, `# Raises` IFF exceptions are raised, `# Arguments` IFF any arguments
- argument order
- presence of a one-line description that starts on the same line as the docstring marker and ends with a dot.

**Comment by Dref360** [comment]
Hey!
I added checks for : 
* blank line before markdown
* order of args
* presence of Markdown
* doesn't throw if the function doesn't have any documentation but is small.
* Argument style name: doc
* Only run on python3

File: keras/backend/common.py [hunk 9]
```diff
@@ -192,6 +191,9 @@ def set_image_dim_ordering(dim_ordering):
         >>> K.image_data_format()
         'channels_last'
     ```
+
+    # Raises
+        ValueError if invalid `'dim_ordering'`
```
Comment (fchollet) [inline keras/backend/common.py]: Since `dim_ordering` is an argument name and not a string variable, quotes are not necessary.

File: keras/backend/tensorflow_backend.py [hunk 10]
```diff
@@ -169,6 +178,17 @@ def set_session(session):
...
+    """Get the type from a string.
+
+    # Arguments
+        dtype: A string representation of a type.
+
+    # Returns:
+        The type requested.
+
+    # Raises
+        ValueError if `'dtype'` is not supported
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Since `dtype` is an argument name and not a string variable, quotes are not necessary.

File: keras/backend/tensorflow_backend.py [hunk 11]
```diff
@@ -190,6 +210,15 @@ def _convert_string_dtype(dtype):
 
 
 def _to_tensor(x, dtype):
+    """Convert the input 'x' to a tensor of type 'dtype'.
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Code delimiters \` would be more appropriate than string quotes here, for `x` and `dtype`.

File: keras/backend/tensorflow_backend.py [hunk 12]
```diff
@@ -773,18 +813,54 @@ def cast(x, dtype):
 
 
 def update(x, new_x):
+    """Update the value of 'x' to 'new_x'.
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Code delimiters ` would be more appropriate than string quotes here, for `x` and `new_x`.

File: keras/backend/tensorflow_backend.py [hunk 13]
```diff
@@ -773,18 +813,54 @@ def cast(x, dtype):
...
+
+    # Arguments
+        x: A Variable.
+        new_x: A tensor of same shape as 'x'.
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Code delimiters ` would be more appropriate than string quotes here, for `x`.

File: keras/backend/tensorflow_backend.py [hunk 14]
```diff
@@ -773,18 +813,54 @@ def cast(x, dtype):
...
+
+    # Returns
+        The variable 'x' updated.
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Code delimiters ` would be more appropriate than string quotes here, for `x`.

File: keras/backend/tensorflow_backend.py [hunk 15]
```diff
@@ -773,18 +813,54 @@ def cast(x, dtype):
...
+    """
     return tf.assign(x, new_x)
 
 
 def update_add(x, increment):
+    """Update the value of 'x' by adding 'increment'.
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Code delimiters ` would be more appropriate than string quotes here, for `x` and `increment`.

File: keras/backend/tensorflow_backend.py [hunk 16]
```diff
@@ -773,18 +813,54 @@ def cast(x, dtype):
...
+
+        # Arguments
+            x: A Variable.
+            increment: A tensor of same shape as 'x'.
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Code delimiters ` would be more appropriate than string quotes here, for `x`.

File: keras/backend/tensorflow_backend.py [hunk 17]
```diff
@@ -773,18 +813,54 @@ def cast(x, dtype):
...
+
+        # Returns
+            The variable 'x' updated.
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Code delimiters ` would be more appropriate than string quotes here, for `x`.

File: keras/backend/tensorflow_backend.py [hunk 18]
```diff
@@ -773,18 +813,54 @@ def cast(x, dtype):
...
+        """
     return tf.assign_add(x, increment)
 
 
 def update_sub(x, decrement):
+    """Update the value of 'x' by substracting 'decrement'.
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Same comment for the rest of the file

File: keras/models.py [hunk 19]
```diff
@@ -293,6 +293,9 @@ def model_from_config(config, custom_objects=None):
 
     # Returns
         A Keras model instance (uncompiled).
+
+    # Raises
+        TypeError if `'config'` is not a dictionary
```
Comment (fchollet) [inline keras/models.py]: Quote characters not required.

File: tests/test_documentation.py [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+        if inspect.isclass(mem):
+            handle_class(name, mem)
+        elif inspect.isfunction(mem):
+            handle_function(name, mem)
+        elif 'keras' in name and inspect.ismodule(mem):
+            # Only test keras' modules
+            handle_module(mem)
+
+
+@pytest.mark.skipif(sys.version_info < (3, 3), reason="requires python3.3")
```
Comment (fchollet) [inline tests/test_documentation.py]: This test should instead have a dedicated command in Travis, see `.travis.yml`. We only want to run it once, and have it have a dedicated status and output.

File: tests/test_documentation.py [hunk 21]
```diff
@@ -0,0 +1,143 @@
...
+@keras_test
```
Comment (fchollet) [inline tests/test_documentation.py]: Not necessary since this test does not instantiate TF graphs.

**Comment by Dref360** [comment]
Done with the changes

**Comment by fchollet** [comment]
Thanks for the work! 👍
